File: misty_jellyfish/bot.py
"""Main bot implementation"""

import asyncio
import logging
import re
import json
import os
from typing import Optional, List, Dict, Any
from atproto import Client, models
import httpx

logger = logging.getLogger(__name__)


class MistyJellyfishBot:
    """Agnostic Bluesky reply bot implementation"""
    
    def __init__(self, handle: str, password: str, config_path: Optional[str] = None):
        self.handle = handle
        self.password = password
        self.client = Client()
        self.running = False
        self.authenticated = False
        self.config = self._load_config(config_path)
        self.compiled_patterns = self._compile_patterns()
    
    def _load_config(self, config_path: Optional[str] = None) -> Dict[str, Any]:
        """Load bot configuration"""
        default_config = {
            "keywords": [],
            "regex_patterns": [],
            "llm_api": {
                "base_url": "http://localhost:1234",
                "model": "local-model",
                "system_prompt": "You are a helpful assistant replying to social media posts. Keep responses brief, friendly, and relevant to the original post.",
                "max_tokens": 100,
                "temperature": 0.7
            },
            "reply_settings": {
                "check_interval": 60,
                "timeline_limit": 20,
                "enable_replies": True
            }
        }
        
        if config_path and os.path.exists(config_path):
            try:
                with open(config_path, 'r') as f:
                    user_config = json.load(f)
                    default_config.update(user_config)
            except Exception as e:
                logger.warning(f"Failed to load config from {config_path}: {e}")
        
        return default_config
    
    def _compile_patterns(self) -> List[re.Pattern]:
        """Compile regex patterns for efficient matching"""
        patterns = []
        for pattern in self.config.get("regex_patterns", []):
            try:
                patterns.append(re.compile(pattern, re.IGNORECASE))
            except re.error as e:
                logger.warning(f"Invalid regex pattern '{pattern}': {e}")
        return patterns
    
    async def start(self):
        """Start the bot"""
        logger.info("Starting reply bot...")
        
        try:
            await self.authenticate()
            self.running = True
            
            interval = self.config["reply_settings"]["check_interval"]
            while self.running:
                await self.monitor_posts()
                await asyncio.sleep(interval)
                
        except Exception as e:
            logger.error(f"Bot error: {e}")
            await self.stop()
    
    async def stop(self):
        """Stop the bot"""
        logger.info("Stopping reply bot...")
        self.running = False
    
    async def authenticate(self):
        """Authenticate with Bluesky"""
        try:
            logger.info(f"Authenticating as {self.handle}...")
            profile = self.client.login(self.handle, self.password)
            self.authenticated = True
            logger.info(f"Successfully authenticated as {profile.display_name or self.handle}")
        except Exception as e:
            logger.error(f"Authentication failed: {e}")
            raise
    
    async def monitor_posts(self):
        """Monitor posts for replies"""
        if not self.authenticated:
            logger.warning("Not authenticated, skipping post monitoring")
            return
            
        if not self.config["reply_settings"]["enable_replies"]:
            logger.debug("Replies disabled in configuration")
            return
            
        try:
            limit = self.config["reply_settings"]["timeline_limit"]
            timeline = self.client.get_timeline(limit=limit)
            
            for post in timeline.feed:
                if self.should_reply_to_post(post):
                    reply_text = await self.generate_reply(post)
                    if reply_text:
                        await self.send_reply(post, reply_text)
                        
        except Exception as e:
            logger.error(f"Error monitoring posts: {e}")
    
    def should_reply_to_post(self, post) -> bool:
        """Determine if we should reply to a post"""
        # Avoid replying to our own posts
        if post.post.author.handle == self.handle:
            return False
        
        # Check if we've already replied (basic check)
        if hasattr(post.post.record, 'reply') and post.post.record.reply:
            return False
            
        post_text = post.post.record.text.lower()
        
        # Check keyword matches
        for keyword in self.config.get("keywords", []):
            if keyword.lower() in post_text:
                logger.debug(f"Keyword match: '{keyword}' in post by {post.post.author.handle}")
                return True
        
        # Check regex pattern matches
        for pattern in self.compiled_patterns:
            if pattern.search(post_text):
                logger.debug(f"Regex match: '{pattern.pattern}' in post by {post.post.author.handle}")
                return True
        
        return False
    
    async def generate_reply(self, post) -> Optional[str]:
        """Generate AI reply using LM Studio API"""
        try:
            post_text = post.post.record.text
            author_handle = post.post.author.handle
            
            # Prepare the prompt
            system_prompt = self.config["llm_api"]["system_prompt"]
            user_prompt = f"Reply to this post by @{author_handle}: \"{post_text}\""
            
            # Call LM Studio API
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{self.config['llm_api']['base_url']}/v1/chat/completions",
                    json={
                        "model": self.config["llm_api"]["model"],
                        "messages": [
                            {"role": "system", "content": system_prompt},
                            {"role": "user", "content": user_prompt}
                        ],
                        "max_tokens": self.config["llm_api"]["max_tokens"],
                        "temperature": self.config["llm_api"]["temperature"]
                    },
                    timeout=30.0
                )
                
                if response.status_code == 200:
                    data = response.json()
                    reply_text = data["choices"][0]["message"]["content"].strip()
                    logger.info(f"Generated reply for @{author_handle}: {reply_text[:50]}...")
                    return reply_text
                else:
                    logger.error(f"LM Studio API error: {response.status_code} - {response.text}")
                    
        except Exception as e:
            logger.error(f"Failed to generate reply: {e}")
        
        return None
    
    async def send_reply(self, original_post, reply_text: str):
        """Send a reply to a post"""
        try:
            # Create reply
            reply_ref = models.create_strong_ref(original_post.post)
            root_ref = original_post.post.record.reply.root if original_post.post.record.reply else reply_ref
            
            self.client.send_post(
                text=reply_text,
                reply_to=models.AppBskyFeedPost.ReplyRef(parent=reply_ref, root=root_ref)
            )
            
            logger.info(f"Sent reply to {original_post.post.author.handle}: {reply_text}")
            
        except Exception as e:
            logger.error(f"Failed to send reply: {e}")

File: internal/bot/bot.go
// Package bot implements the MISTY JELLYFISH Bluesky reply bot.
package bot

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"regexp"
	"strings"
	"time"

	"github.com/tigwyk/mistyjellyfish/internal/config"
)

const defaultBskyHost = "https://bsky.social"

// session holds the authenticated AT Protocol session tokens.
type session struct {
	AccessJWT  string `json:"accessJwt"`
	RefreshJWT string `json:"refreshJwt"`
	Handle     string `json:"handle"`
	DID        string `json:"did"`
}

// StrongRef is an AT Protocol strong reference (cid + uri).
type StrongRef struct {
	CID string `json:"cid"`
	URI string `json:"uri"`
}

// ReplyRef is the reply field of an AppBskyFeedPost record.
type ReplyRef struct {
	Root   StrongRef `json:"root"`
	Parent StrongRef `json:"parent"`
}

// feedPost is the record type for a Bluesky post.
type feedPost struct {
	Type      string    `json:"$type"`
	Text      string    `json:"text"`
	Reply     *ReplyRef `json:"reply,omitempty"`
	CreatedAt string    `json:"createdAt"`
}

// timelineResponse is the API response from app.bsky.feed.getTimeline.
type timelineResponse struct {
	Feed []FeedViewPost `json:"feed"`
}

// FeedViewPost wraps a post in the timeline feed.
type FeedViewPost struct {
	Post PostView `json:"post"`
}

// PostView holds post metadata and the record.
type PostView struct {
	URI    string     `json:"uri"`
	CID    string     `json:"cid"`
	Author AuthorView `json:"author"`
	Record FeedRecord `json:"record"`
}

// AuthorView holds author information.
type AuthorView struct {
	DID         string `json:"did"`
	Handle      string `json:"handle"`
	DisplayName string `json:"displayName"`
}

// FeedRecord is the raw record JSON of a post, including optional reply field.
type FeedRecord struct {
	Type  string    `json:"$type"`
	Text  string    `json:"text"`
	Reply *ReplyRef `json:"reply,omitempty"`
}

// chatMessage is a message for the LLM API.
type chatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// chatRequest is the request body for the OpenAI-compatible chat API.
type chatRequest struct {
	Model       string        `json:"model"`
	Messages    []chatMessage `json:"messages"`
	MaxTokens   int           `json:"max_tokens"`
	Temperature float64       `json:"temperature"`
}

// chatResponse is the response body from the OpenAI-compatible chat API.
type chatResponse struct {
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	} `json:"choices"`
}

// Bot is the MISTY JELLYFISH Bluesky reply bot.
type Bot struct {
	handle   string
	password string
	config   config.BotConfig
	patterns []*regexp.Regexp
	sess     *session
	client   *http.Client
	bskyHost string
}

// New creates a Bot from the supplied credentials and configuration,
// connecting to the default Bluesky host (https://bsky.social).
func New(handle, password string, cfg config.BotConfig) (*Bot, error) {
	return NewWithHost(defaultBskyHost, handle, password, cfg)
}

// NewWithHost creates a Bot like New but with a configurable Bluesky host URL.
// This is useful for testing against a local httptest.Server.
func NewWithHost(bskyHost, handle, password string, cfg config.BotConfig) (*Bot, error) {
	patterns, err := compilePatterns(cfg.RegexPatterns)
	if err != nil {
		return nil, err
	}

	return &Bot{
		handle:   handle,
		password: password,
		config:   cfg,
		patterns: patterns,
		client:   &http.Client{Timeout: 30 * time.Second},
		bskyHost: bskyHost,
	}, nil
}

// MakeFeedViewPost constructs a FeedViewPost for use in tests.
func MakeFeedViewPost(handle, text, uri, cid string, reply *ReplyRef) FeedViewPost {
	return FeedViewPost{
		Post: PostView{
			URI: uri,
			CID: cid,
			Author: AuthorView{
				Handle: handle,
			},
			Record: FeedRecord{
				Type:  "app.bsky.feed.post",
				Text:  text,
				Reply: reply,
			},
		},
	}
}

// compilePatterns compiles regex patterns, logging and skipping invalid ones.
func compilePatterns(raw []string) ([]*regexp.Regexp, error) {
	out := make([]*regexp.Regexp, 0, len(raw))
	for _, p := range raw {
		compiled, err := regexp.Compile("(?i)" + p)
		if err != nil {
			log.Printf("Warning: invalid regex pattern %q: %v", p, err)
			continue
		}
		out = append(out, compiled)
	}
	return out, nil
}

// Run starts the bot loop. It blocks until ctx is cancelled.
func (b *Bot) Run(ctx context.Context) error {
	log.Println("Starting MISTY JELLYFISH reply bot...")

	if err := b.Authenticate(ctx); err != nil {
		return fmt.Errorf("authentication failed: %w", err)
	}

	interval := time.Duration(b.config.ReplySettings.CheckInterval) * time.Second
	ticker := time.NewTicker(interval)
	defer ticker.Stop()

	// Run once immediately, then on each tick.
	b.MonitorPosts(ctx)

	for {
		select {
		case <-ctx.Done():
			log.Println("Stopping reply bot...")
			return nil
		case <-ticker.C:
			b.MonitorPosts(ctx)
		}
	}
}

// Authenticate creates an AT Protocol session.
func (b *Bot) Authenticate(ctx context.Context) error {
	log.Printf("Authenticating as %s...", b.handle)

	body, _ := json.Marshal(map[string]string{
		"identifier": b.handle,
		"password":   b.password,
	})

	resp, err := b.xrpcPost(ctx, "/xrpc/com.atproto.server.createSession", "", body)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		raw, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("createSession returned %d: %s", resp.StatusCode, raw)
	}

	var s session
	if err := json.NewDecoder(resp.Body).Decode(&s); err != nil {
		return fmt.Errorf("decoding session response: %w", err)
	}

	b.sess = &s
	log.Printf("Authenticated as %s (DID: %s)", s.Handle, s.DID)
	return nil
}

// MonitorPosts fetches the timeline and processes matching posts.
func (b *Bot) MonitorPosts(ctx context.Context) {
	if b.sess == nil {
		log.Println("Not authenticated, skipping post monitoring")
		return
	}

	if !b.config.ReplySettings.EnableReplies {
		log.Println("Replies disabled in configuration")
		return
	}

	timeline, err := b.getTimeline(ctx)
	if err != nil {
		log.Printf("Error fetching timeline: %v", err)
		return
	}

	for _, item := range timeline.Feed {
		if !b.ShouldReplyTo(item) {
			continue
		}

		replyText, err := b.GenerateReply(ctx, item)
		if err != nil {
			log.Printf("Error generating reply: %v", err)
			continue
		}
		if replyText == "" {
			continue
		}

		if err := b.sendReply(ctx, item, replyText); err != nil {
			log.Printf("Error sending reply: %v", err)
		}
	}
}

// getTimeline fetches the authenticated user's home timeline.
func (b *Bot) getTimeline(ctx context.Context) (*timelineResponse, error) {
	url := fmt.Sprintf("%s/xrpc/app.bsky.feed.getTimeline?limit=%d",
		b.bskyHost, b.config.ReplySettings.TimelineLimit)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+b.sess.AccessJWT)

	resp, err := b.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		raw, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("getTimeline returned %d: %s", resp.StatusCode, raw)
	}

	var tl timelineResponse
	if err := json.NewDecoder(resp.Body).Decode(&tl); err != nil {
		return nil, fmt.Errorf("decoding timeline response: %w", err)
	}
	return &tl, nil
}

// ShouldReplyTo returns true when the bot should reply to the given post.
func (b *Bot) ShouldReplyTo(item FeedViewPost) bool {
	// Skip our own posts.
	if item.Post.Author.Handle == b.handle {
		return false
	}

	// Skip posts that are already replies (to keep things tidy).
	if item.Post.Record.Reply != nil {
		return false
	}

	text := strings.ToLower(item.Post.Record.Text)

	for _, kw := range b.config.Keywords {
		if strings.Contains(text, strings.ToLower(kw)) {
			log.Printf("Keyword match %q in post by @%s", kw, item.Post.Author.Handle)
			return true
		}
	}

	for _, p := range b.patterns {
		if p.MatchString(text) {
			log.Printf("Regex match %q in post by @%s", p.String(), item.Post.Author.Handle)
			return true
		}
	}

	return false
}

// GenerateReply calls the LM Studio (OpenAI-compatible) API and returns the reply text.
func (b *Bot) GenerateReply(ctx context.Context, item FeedViewPost) (string, error) {
	author := item.Post.Author.Handle
	postText := item.Post.Record.Text

	payload := chatRequest{
		Model: b.config.LLMAPI.Model,
		Messages: []chatMessage{
			{Role: "system", Content: b.config.LLMAPI.SystemPrompt},
			{Role: "user", Content: fmt.Sprintf("Reply to this post by @%s: %q", author, postText)},
		},
		MaxTokens:   b.config.LLMAPI.MaxTokens,
		Temperature: b.config.LLMAPI.Temperature,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}

	url := b.config.LLMAPI.BaseURL + "/v1/chat/completions"
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := b.client.Do(req)
	if err != nil {
		return "", fmt.Errorf("calling LLM API: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		raw, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("LLM API returned %d: %s", resp.StatusCode, raw)
	}

	var cr chatResponse
	if err := json.NewDecoder(resp.Body).Decode(&cr); err != nil {
		return "", fmt.Errorf("decoding LLM response: %w", err)
	}

	if len(cr.Choices) == 0 {
		return "", fmt.Errorf("LLM API returned no choices")
	}

	replyText := strings.TrimSpace(cr.Choices[0].Message.Content)
	log.Printf("Generated reply for @%s: %.50s...", author, replyText)
	return replyText, nil
}

// sendReply posts a reply to the given post.
func (b *Bot) sendReply(ctx context.Context, item FeedViewPost, replyText string) error {
	parent := StrongRef{CID: item.Post.CID, URI: item.Post.URI}

	// Determine root: if the original post has a reply chain, use its root.
	root := parent
	if item.Post.Record.Reply != nil {
		root = item.Post.Record.Reply.Root
	}

	record := feedPost{
		Type:      "app.bsky.feed.post",
		Text:      replyText,
		Reply:     &ReplyRef{Root: root, Parent: parent},
		CreatedAt: time.Now().UTC().Format(time.RFC3339),
	}

	payload := map[string]interface{}{
		"repo":       b.sess.DID,
		"collection": "app.bsky.feed.post",
		"record":     record,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	resp, err := b.xrpcPost(ctx, "/xrpc/com.atproto.repo.createRecord", b.sess.AccessJWT, body)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		raw, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("createRecord returned %d: %s", resp.StatusCode, raw)
	}

	log.Printf("Sent reply to @%s: %s", item.Post.Author.Handle, replyText)
	return nil
}

// xrpcPost is a helper that sends a JSON POST to the Bluesky XRPC endpoint.
func (b *Bot) xrpcPost(ctx context.Context, path, accessJWT string, body []byte) (*http.Response, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		b.bskyHost+path, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	if accessJWT != "" {
		req.Header.Set("Authorization", "Bearer "+accessJWT)
	}
	return b.client.Do(req)
}
